source/common/router/router.cc

```diff
@@ -530,6 +561,16 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
                                          grpc_request_, hedging_params_.hedge_on_per_try_timeout_,
                                          config_.respect_expected_rq_timeout_);
 
+  if (route_entry_->measureTimeoutOnRequestStart() && timeout_.global_timeout_.count() > 0) {
+    // if we measure the timeout from the request start, calculate the deadline for the request
+    // so we can update timeout headers just before the request is sent upstream.
+    Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+    deadline_ = dispatcher.timeSource().monotonicTime() + timeout_.global_timeout_;
+    // start measuring timeout.
+    response_timeout_ = dispatcher.createTimer([this]() -> void { onResponseTimeout(); });
+    response_timeout_->enableTimer(timeout_.global_timeout_);
```

Comment:
I would make a common method for starting the response timeout timer for clarity.